contents/Documentation/FP concepts.playground/Pages/Type classes.xcplaygroundpage/Contents.swift

```diff
@@ -4,11 +4,144 @@
+ */
+protocol Transformer {
+    static func transform<A, B>(_ fa: Kind<Self, A>, _ f: (A) -> B) -> Kind<Self, B>
+}
+/*:
+ ## Type class laws
+
+ As we mentioned above, type classes are governed by **algebraic laws**. This means that all implementations of the type class must behave on a certain and satisfy some properties. Each type class has its own laws, as the ones we presented for `Equatable` above.
+
+ For the `Transformer` type class that we are defining, we can consider some laws as well:
+
+ - **Identity**: if we transform using the identity function (`id`), we must obtain the original value; i.e., `F.transform(fa, id) == fa`.
+ - **Composition**: if we apply two transformations to a value, we must obtain the same result as if we apply a single transformation with the composition of the two transforming functions; i.e., `F.transform(F.transform(fa, f), g) == F.transform(fa, g <<< f)`.
+
+ Swift does not provide a way of encoding these laws and enforcing them in every implementation. To do this, Bow type classes encode their laws as property-based tests using [SwiftCheck](https://github.com/typelift/SwiftCheck). Then, each implementation can be tested against these laws to guarantee that the implementation satisfies them. Type class laws help us have a better reasoning about our code and some times help us rewrite pieces for optimization (like in the case of `Transformer` where we can compose functions and apply a single transformation instead of two).
+
+ Notice that sometimes you can provide an implementation for a given type class that does not satisfies the laws. Those are called **lawless instances**, in contrast to the ones that satisfy them, which are called **lawful instances**.
+
+ ## Type class instances
+
+ A **type class instance** is a concrete implementation of a type class for a given type. Instances are usually created through the extension mechanisms that Swift provides. For instance, let us provide an instance of `Transformer` for the `Option` data type in Bow:
+ */
+// nef:begin:hidden
+postfix operator ^
+postfix func ^<A>(_ value: OptionOf<A>) -> Option<A> {
+    return Option.fix(value)
+}
+// nef:end
+extension ForOption: Transformer {
+    static func transform<A, B>(_ fa: Kind<ForOption, A>, _ f: (A) -> B) -> Kind<ForOption, B> {
+        return fa^.fold(Option<B>.none,                // It is empty, no transformation
+                        { a in Option<B>.some(f(a)) }) // Transformed with f and wrapped in an Option<B>
+    }
+}
+/*:
+ Notice that, since the type class works at the kind level, it is not an extension of `Option`, but and extension of `ForOption`. Revisiting the definition of `Transformer`, the `Self` requirement is used in `Kind<Self, A>`. If we disect the definition of `Option<A>`, it extends `Kind<ForOption, A>`. By doing some sort of pattern matching between `Kind<Self, A>` and `Kind<ForOption, A>`, we can see that it must be `ForOption` the candidate to do the extension.
```

Comment:
```suggestion
 Notice that, since the type class works at the kind level, it is not an extension of `Option`, but an extension of `ForOption`. Revisiting the definition of `Transformer`, the `Self` requirement is used in `Kind<Self, A>`. If we dissect the definition of `Option<A>`, it extends `Kind<ForOption, A>`. By doing some sort of pattern matching between `Kind<Self, A>` and `Kind<ForOption, A>`, we can see that it must be `ForOption` the candidate to do the extension.
```